File: leetcode_ly/c++/Tree/BinaryTreePath.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        findPath(ans,root,"");
        return ans;
    }
    
    void findPath(vector<string>& ans, TreeNode* root, string s){
        if(root == NULL) return;
        if(root->left == NULL && root->right == NULL){
            ans.push_back(s + to_string(root->val));
            return;
        }
        else{
            findPath(ans,root->left,s + to_string(root->val) + "->");
            findPath(ans,root->right,s + to_string(root->val) + "->");
        }
    }
};

File: leetcode_ly/java/Tree/BinaryTreePath.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> res = new ArrayList<String>();
        if(root == null) return res;
        dfs(res,root,"" + root.val);
        return res;
    }
    
    public void dfs(List<String> res,TreeNode root,String s){
        if(root.left == null && root.right == null){
            res.add(s);
            return;
        }
        if(root.left != null) dfs(res,root.left,s + "->" + root.left.val);
        if(root.right != null) dfs(res,root.right,s + "->" + root.right.val);
    }
}